src/libraries/trading/PlaceOrder.sol

```diff
@@ -0,0 +1,135 @@
+// price is the exact price you want to buy/sell at [which may not be the cost, for example to short a binary market it'll cost 1-price, to go long it'll cost price]
+// smallest order value is 10**14 WEI
+
+pragma solidity ^0.4.13;
+
+import 'ROOT/IController.sol';
+import 'ROOT/libraries/math/SafeMathUint256.sol';
+import 'ROOT/trading/Trading.sol';
+import 'ROOT/trading/IOrders.sol';
+
+
+library PlaceOrder {
+    using SafeMathUint256 for uint256;
+
+    uint256 constant minOrderValue = 10**14;
+
+    struct Data {
+        // Contracts
+        IOrders orders;
+        IMarket market;
+
+        // Order
+        address sender;
+        uint8 outcome;
+        Trading.TradeTypes tradeType;
+        uint256 attoshares;
+        int256 displayPrice;
+        bytes32 betterOrderId;
+        bytes32 worseOrderId;
+        uint256 tradeGroupId;
+    }
+
+    //
+    // Constructor
+    //
+
+    function create(IController _controller, address _sender, uint8 _outcome, Trading.TradeTypes _type, uint256 _attoshares, int256 _displayPrice, IMarket _market, bytes32 _betterOrderId, bytes32 _worseOrderId, uint256 _tradeGroupId) internal returns (Data) {
+        require(_market.getTypeName() == "Market");
+        require(_outcome < _market.getNumberOfOutcomes());
+        require(_market.getMinDisplayPrice() < _displayPrice && _displayPrice < _market.getMaxDisplayPrice());
+
+        IOrders _orders = IOrders(_controller.lookup("Orders"));
+
+        return Data({
+            orders: _orders,
+            market: _market,
+            sender: _sender,
+            outcome: _outcome,
+            tradeType: _type,
+            attoshares: _attoshares,
+            displayPrice: _displayPrice,
+            betterOrderId: _betterOrderId,
+            worseOrderId: _worseOrderId,
+            tradeGroupId: _tradeGroupId
+        });
+    }
+
+    //
+    // "public" functions
+    //
+
+    function placeBid(PlaceOrder.Data _data) internal returns (uint256, uint256) {
+        uint256 _orderValueInAttotokens = SafeMathUint256.fxpMul(_data.attoshares, SafeMathUint256.sub(_data.market.getCompleteSetCostInAttotokens(), uint256(_data.displayPrice)), 1 ether);
+        require(_orderValueInAttotokens >= minOrderValue);
+
+        uint256 _attotokensEscrowed = 0;
+        uint256 _attosharesEscrowed = 0;
+        uint256 _attosharesToCover = _data.attoshares;
+        uint8 _numberOfOutcomes = _data.market.getNumberOfOutcomes();
+
+        // Figure out how many almost-complete-sets (just missing `outcome` share) the maker has
+        uint256 _attosharesHeld = 2**254;
+        uint8 _i = 0;
+        while (_i < _numberOfOutcomes) {
```

Comment:
This should be a for loop (Serpent didn't have for loops which is why everything was a while loop).

Reply:
Same comment for most other while loops.

Reply:
Ok, sounds good.  I'll change this to a for loop.